Run workflow updates for saved cloud plans (#433)
hashicorp/terraform-docs-common: website/docs/cloud-docs/run/modes-and-options.mdx

Terraform Cloud runs support many of the same modes and options available in the Terraform CLI.

## Plan and Apply (Standard)

The default run mode of Terraform Cloud is to perform a plan, then apply it. If auto-apply is enabled, a successful plan will apply immediately; otherwise the run will wait for user confirmation before applying.

- **CLI:** Use `terraform apply` (without providing a saved plan file).

> - **API:** Create a run without specifying any options that would select a different mode.

Reviewer: ```suggestion
- **API:** Create a run without specifying any options that select a different mode.
```

- **API:** Use the `is-destroy` option.
- **UI:** Use a workspace's **Destruction and Deletion** settings page.

## Plan Only/Speculative Plan

> This option creates a [speculative plan](/terraform/cloud-docs/run/remote-operations#speculative-plans). The speculative plan shows a set of possible changes and checks them against Sentinel policies, but Terraform cannot apply it.

Reviewer: ```suggestion
This option creates a [speculative plan](/terraform/cloud-docs/run/remote-operations#speculative-plans). The speculative plan shows a set of possible changes and checks them against Sentinel policies, but Terraform can _not_ apply this plan.
```

> You can create speculative plans with a Terraform version that is different from the one currently selected for the workspace. This lets you check whether your configuration is compatible with a newer Terraform version without changing the workspace settings.

Reviewer: ```suggestion
You can create speculative plans with a different Terraform version than the one currently selected for a workspace. This lets you check whether your configuration is compatible with a newer Terraform version without changing the workspace settings.
```

> Plan-only runs ignore the per-workspace run queue; they can proceed even if another run is in progress, they cannot become the workspace's current run, and they do not block progress on other runs in the workspace.

Reviewer: ```suggestion
Plan-only runs ignore the per-workspace run queue. Plan-only runs can proceed even if another run is in progress, can not become the workspace's current run, and do not block progress on a workspace's other runs.
```

- **API:** Set the `plan-only` option to `true` and specify an available terraform version using the `terraform-version` field.
- **UI:** Open the workspace's **Actions** menu, select **Start new run**, and then choose **Plan only** as the run type.

> - **VCS:** When a workspace is connected to a VCS repository, Terraform Cloud automatically starts a speculative plan when a pull request (or merge request) is opened against the selected branch of that repository. The plan will be linked from the pull/merge request view in your VCS, and can also be found in the workspace's run list.

Reviewer: ```suggestion
- **VCS:** When a workspace is connected to a VCS repository, Terraform Cloud automatically starts a speculative plan when someone opens a pull request (or merge request) against the selected branch of that repository.  The pull/merge request view in your VCS links to the speculative plan, and you can also find it in the workspace's run list.
```

## Saved Plans

-> **Version note:** Using saved plans from the CLI with Terraform Cloud requires at least Terraform CLI v1.6.0.

> Saved plan runs are very similar to standard plan and apply runs: they perform a plan, and then optionally apply it. There are three main differences:

Reviewer: ```suggestion
Saved plan runs are very similar to standard plan and apply runs: they perform a plan and then optionally apply it. There are three main differences:
```

> 1. _No wait for planning._ Saved plan runs ignore the per-workspace run queue during their plan and checks; like plan-only runs, they can begin planning even if another run is in progress, without blocking progress on other runs. They only become the current run (and block other runs) once they are confirmed for apply.

Reviewer: I think we can omit the last sentence because you cover that in the below point:
```suggestion
1. _No wait for planning._ Saved plan runs ignore the per-workspace run queue during their plan and checks. Like plan-only runs, saved plans can begin planning even if another run is in progress, without blocking progress on other runs.
```

> 2. _No auto-apply._ Saved plan runs are never auto-applied, even if auto-apply is enabled for the workspace; they will only apply when confirmed by a user.

Reviewer: ```suggestion
2. _No auto-apply._ Saved plan runs are never auto-applied, even if you enabled auto-apply for the workspace. Saved plans only apply if you confirm them.
```

> 3. _Automatic discard for stale plans._ If another run is applied (or the state is otherwise modified) before a saved plan run is confirmed, Terraform Cloud will automatically discard that saved plan, since it would no longer make sense to apply it. Saved plans may also be automatically discarded if they are not confirmed within a few weeks.

Reviewer: Should we tell folks how to automatically discard saved plans? Or does Terraform Cloud do that without folks having to enable it?
```suggestion
3. _Automatic discard for stale plans._ If another run is applied (or the state is otherwise modified) before a saved plan run is confirmed, Terraform Cloud automatically discards that saved plan. You can also automatically discard saved plans if you do not are confirm them within a few weeks.
```

Author: That's default behavior that you cannot opt out of.

Author: ```suggestion
3. _Automatic discard for stale plans._ If another run is applied (or the state is otherwise modified) before a saved plan run is confirmed, Terraform Cloud automatically discards that saved plan. Terraform Cloud may also automatically discard saved plans if they are not confirmed within a few weeks.
```

> Saved plans are ideal for interactive CLI workflows (where you might perform many exploratory plans and then choose one to apply), or for custom continuous integration workflows where the default run queue behavior isn't suitable.

Reviewer: Do we need to expand on why the default run behavior isn't suitable for continuous integration workflows? Or will folks understand inherently?
```suggestion
Saved plans are ideal for interactive CLI workflows, where you can perform many exploratory plans and then choose one to apply, or for custom continuous integration workflows where the default run queue behavior isn't suitable.
```

Author: I kind of file this one under "if you need it, you know you need it". The short answer is that it's the same reasons why you'd use saved plans _without_ TFC; the long answer would be distracting here.

> - **CLI:** Use `terraform plan -out <FILE>` to perform and save a plan, then use `terraform apply <FILE>` to apply the saved plan. Use `terraform show <FILE>` if you need to inspect a saved plan before applying it.

Reviewer: ```suggestion
- **CLI:** Use `terraform plan -out <FILE>` to perform and save a plan, then use `terraform apply <FILE>` to apply the saved plan. Use `terraform show <FILE>` to inspect a saved plan before applying it.
```
